packages/enclave-contracts/contracts/Enclave.sol

```diff
@@ -296,7 +302,7 @@ contract Enclave is IEnclave, OwnableUpgradeable {
         uint256 e3Id,
         bytes calldata plaintextOutput,
         bytes calldata proof
-    ) external returns (bool success) {
+    ) external nonReentrant returns (bool success) {
```

Comment:
I would suggest not using the guard here and instead just following the checks-effect-interaction patter. That would achieve the same result but not incur any costs to the users.

i.e. move 
`e3s[e3Id].ciphertextOutput = ciphertextOutputHash;` before the external call to `e3Program.verify` and even if e3Program was a malicious contract and re-entered, there would be nothing they could do. Furthermore, e3 programs are whitelisted using `enableE3Program`, thus are not arbitrary external contracts. 

The same logic should be applied to `publishPlaintextOutput` -> move the state update `e3s[e3Id].plaintextOutput = plaintextOutput;` before the external call to the decryption verifier

Reply:
Yep, you’re right, I’ll remove the unnecessary parts. My reasoning was: “let’s add the guard” as an extra security layer. But if we apply the CEI pattern correctly, we can remove the guards and save some gas. The important part is making sure the CEI pattern continues to be followed consistently in the future.

Same for the other review.